### PR TITLE
Add Gitlab retriever

### LINKED ISSUE #695: (Gitlab Retriever) Add logic to allow gitlab retriever and non-public Git Repos
### Requirements

We'd use this project for feature flags but need to use a private Gitlabd repo to host our flags.yaml rules.

I'll create a PR shortly with a suggested code addition for the retriever's folder.

### PR DESCRIPTION
**Pull request by ruairi-wmf**
# Description

# Changes include
- [ ] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking changes (change that is not backward-compatible and/or changes current functionality)

# Closes issue(s)
Resolve #695

# Checklist
- [X] I have tested this code
- [X] I have added unit test to cover this code
- [ ] I have updated the documentation (`README.md` and `/website/docs`)
- [X] I have followed the [contributing guide](CONTRIBUTING.md)

### CODE REVIEW

**Review by thomaspoignant — CHANGES_REQUESTED** [review]
Hello @ruairi-wmf thanks a lot for your pull request this is an awesome addition to GO Feature Flag.

I have looked at your PR and I have a few things that are not working.
I saw that you are not using the GitLab API to retrieve the files but the public raw access from the website.

From my early investigation, I think it could be great to use the API instead.
This API can help us doing what we want to do: https://docs.gitlab.com/ee/api/repository_files.html#get-raw-file-from-repository

I've done a quick CURL to test it and it seems to work perfectly.

```shell
curl --header "PRIVATE-TOKEN: <GITLAB TOKEN>" "https://gitlab.com/api/v4/projects/<REPOSITORY SLUG url encoded>/repository/files/<PATH OF YOUR file url encoded>/raw?ref=main"
```

A point of attention is to be sure that you URL encode both the repository slug and the path of your file.
- If your slug is `thomas.poignant/test-go-feature-flag-provider` is should be `thomas.poignant%2Ftest-go-feature-flag-provider` in the URL.
- If your path is `testdata/flags.yaml` it should be `testdata%2Fflags.yaml` in the URL.

**Comment by ruairi-wmf** [comment]
Thanks Thomas, I'll work on this tonight.  Cheers!

File: retriever/gitlabretriever/retriever.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+		return nil, fmt.Errorf("missing mandatory information URL=%s, FilePath=%s", r.URL, r.FilePath)
+	}
+
+	// default branch is main
+	branch := r.Branch
+	if branch == "" {
+		branch = "main"
+	}
+
+	// add header for Github Token if specified
```
[the remark this answers is not in the document]
Comment (ruairi-wmf) [reply retriever/gitlabretriever/retriever.go:38]: Fixed

File: cmd/relayproxy/config/retriever.go [hunk 2]
```diff
@@ -73,7 +78,7 @@ const (
 // IsValid is checking if the value is part of the enum
 func (r RetrieverKind) IsValid() error {
 	switch r {
-	case HTTPRetriever, GitHubRetriever, S3Retriever, FileRetriever, GoogleStorageRetriever, KubernetesRetriever:
+	case HTTPRetriever, GitHubRetriever, GitlabRetriever, S3Retriever, FileRetriever, GoogleStorageRetriever, KubernetesRetriever:
```
[the remark this answers is not in the document]
Comment (ruairi-wmf) [reply cmd/relayproxy/config/retriever.go:81]: Fixed

File: retriever/gitlabretriever/retriever.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	header := http.Header{}
+	if r.GitlabToken != "" {
+		header.Add("Authorization", fmt.Sprintf("token %s", r.GitlabToken))
+	}
+
+	URL := filepath.Join(r.URL, "-raw", branch, r.FilePath)
```
[the remark this answers is not in the document]
Comment (ruairi-wmf) [reply retriever/gitlabretriever/retriever.go:44]: Fixed

File: cmd/relayproxy/service/gofeatureflag_test.go [hunk 4]
```diff
@@ -44,6 +46,23 @@ func Test_initRetriever(t *testing.T) {
 				Timeout:        20 * time.Millisecond,
 			},
 		},
+		{
```
Comment (thomaspoignant) [inline cmd/relayproxy/service/gofeatureflag_test.go:49]: Can you add another test to validate that it works without setting the branch name?

File: retriever/gitlabretriever/retriever.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+func (r *Retriever) Retrieve(ctx context.Context) ([]byte, error) {
+	if r.FilePath == "" || r.URL == "" {
+		return nil, fmt.Errorf("missing mandatory information URL=%s, FilePath=%s", r.URL, r.FilePath)
+	}
+
+	// default branch is main
+	branch := r.Branch
+	if branch == "" {
+		branch = "main"
+	}
```
Comment (thomaspoignant) [inline retriever/gitlabretriever/retriever.go:38]: The variable `branch` is not used.

File: retriever/gitlabretriever/retriever.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	// add header for Gitlab Token if specified
+	header := http.Header{}
+	if r.GitlabToken != "" {
+		header.Add("PRIVATE-TOKEN", r.GitlabToken)
+	}
+
+	URL = strings.Trim(r.URL, "/")
+	slug := strings.Trim(r.RepositorySlug, "/")
+	path := strings.Trim(r.FilePath, "/")
+	reqURL := fmt.Sprintf("%s/api/v4/projects/%s/repository/files/%s/raw?ref=%s", URL, url.QueryEscape(slug), url.QueryEscape(path), r.Branch)
```
Comment (thomaspoignant) [inline retriever/gitlabretriever/retriever.go:54]:
```suggestion
	reqURL := fmt.Sprintf("%s/api/v4/projects/%s/repository/files/%s/raw?ref=%s", URL, url.QueryEscape(slug), url.QueryEscape(path), branch)
```

File: retriever/gitlabretriever/retriever.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	// default branch is main
+	branch := r.Branch
+	if branch == "" {
+		branch = "main"
+	}
+
+	URL := r.URL
+	if URL == "" {
+		URL = "https://gitlab.com"
+	}
```
Comment (thomaspoignant) [inline retriever/gitlabretriever/retriever.go:43]: This is not used, `URL` is re-affected in line 51.

File: retriever/gitlabretriever/retriever_test.go [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+		// {
+		// 	name: "HTTP Error",
+		// 	fields: fields{
+		// 		httpClient:     mock.HTTP{},
+		// 		URL:            "https://gitlab.com/error",
+		// 		RepositorySlug: "aa/go-feature-flags-config",
+		// 		filePath:       "bad-file/file.yaml",
+		// 	},
+		// 	wantErr: true,
+		// },
```
Comment (thomaspoignant) [inline retriever/gitlabretriever/retriever_test.go:88]: Should we uncomment this?

File: retriever/gitlabretriever/retriever.go [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+	"github.com/thomaspoignant/go-feature-flag/internal"
+)
+
+// Retriever is a configuration struct for a GitHub retriever.
+type Retriever struct {
+	Branch         string // default is main
+	FilePath       string
+	GitlabToken    string
+	RepositorySlug string
+	URL            string        // https://gitlab.com
```
Comment (thomaspoignant) [inline retriever/gitlabretriever/retriever.go:22]: If I understand correctly URL is the based DNS of your gitlab instance?
If yes I am not sure if the URL is the right name here.

I would use something like GitlabBaseURL.

WDYT?

File: retriever/gitlabretriever/retriever.go [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+		URL = "https://gitlab.com"
+	}
+
+	// add header for Gitlab Token if specified
+	header := http.Header{}
+	if r.GitlabToken != "" {
+		header.Add("PRIVATE-TOKEN", r.GitlabToken)
+	}
+
+	URL = strings.Trim(r.URL, "/")
```
Comment (thomaspoignant) [inline retriever/gitlabretriever/retriever.go:51]: Would it be better to use `url.Parse` (https://pkg.go.dev/net/url#Parse) and extract the base DNS (`url.Host`) and the scheme (`url.Scheme`).

**Review by thomaspoignant — CHANGES_REQUESTED** [review]
Hey thanks again for the changes, I have left some new comments.
If you need any help or more information let me know.

**Comment by thomaspoignant** [comment]
Hey @ruairi-wmf I have used your PR to finish it with my remarks.
I will merge it Today and create a new version `1.10.0`, you will be able to use the Gitlab retriever starting this version.

**Review by thomaspoignant — APPROVED** [review]

**Pull request merged**